article_store/articles/api.py

```diff
@@ -0,0 +1,206 @@
+
+    return root
+
+
+class ArticleXMLAPIView(APIView):
+    parser_classes = (ArticleXMLParser,)
+    renderer_classes = (XMLRenderer,)
+
+    default_part_name = 'front'
+    latest = 'latest'
+    style_element = '<?xml version="1.0" ?>\n'
+
+    def post(self, request: Request, article_id: str) -> HttpResponse:
+        """
+
+        :param request: class: `Request`
+        :param article_id: str
+        :return: class: `HttpResponse`
+        """
+
+        if not article_id or not Article.id_is_valid(article_id):
+            return Response({'error': 'Please provide a valid article id'}, status=status.HTTP_406_NOT_ACCEPTABLE)
+
+        article, created = Article.objects.get_or_create(id=article_id)
+        article_version = ArticleVersion.objects.create(version=article.next_version,
+                                                        article=article)
+
+        for content_item in request.data.get('content-list', []):
+            Content.objects.create(article_version=article_version, **content_item)
+
+        serializer = ArticleVersionSerializer(article_version)
+
+        return Response(serializer.data, status=status.HTTP_201_CREATED)
+
+    @staticmethod
+    def put(request: Request, article_id: str, version: str) -> HttpResponse:
+        """Update an `ArticleVersion`
+
+        :param request:
+        :param article_id:
+        :param version:
+        :return: class: `HttpResponse`
+        """
+        if article_id and version:
+            article_version = ArticleVersion.objects.get(version=version, article_id=article_id)
+
+            old_content = Content.objects.filter(article_version=article_version)
+            old_content.delete()
+
+            for content_item in request.data.get('content-list', []):
+                Content.objects.create(article_version=article_version, **content_item)
+
+            serializer = ArticleVersionSerializer(article_version)
+
+            return Response(serializer.data, status=status.HTTP_201_CREATED)
+
+        return Response({'error': 'Please provide a valid article id and version number'},
+                        status=status.HTTP_406_NOT_ACCEPTABLE)
+
+    def delete(self, request: Request, article_id: str, version: str = '') -> HttpResponse:
+        """Delete an `Article` or a all versions including and > a specified `ArticleVersion`
+
+        :param request: class: `Request`
+        :param article_id: str
+        :param version: str
+        :return: class: `HttpResponse`
+        """
+        try:
+            article = Article.objects.get(id=article_id)
+
+            if not version:
+                article.delete()
+            else:
+                if version == self.latest:
+                    version = article.latest_version
+
+                article_versions = ArticleVersion.objects.filter(article_id=article_id)
+
+                for article_version in article_versions:
+                    if article_version.version >= int(version):
```

Comment:
if I delete version 2, then version 3 is deleted as well: no opinion on that, but it leaves the user freedom. At eLife we practically never delete anything, only unpublished versions ingested by error.

Reply:
(freedom as opposed to: refusing to delete version 2 if a version 3 exists)